### PR TITLE
Make the Docker Sandbox Stateful (e.g., keep track of all `cd`ed directories)

### PR DESCRIPTION
**Pull request by xingyaoww**
This PR tries to resolve the issue discussed [here](https://github.com/OpenDevin/OpenDevin/issues/226#issuecomment-2021943071) by making the Docker Sandbox Stateful (e.g., keeping track of all `cd`ed directories).

The current draft solution (not working!) is built on [`dockerpty`](https://github.com/d11wtq/dockerpty/blob/master/dockerpty/pty.py), but i feel it is kinda tricky (may need more tweaks) to get it working since I'm not an expert in PTY :-(

Will try to look into this in the next few days -- suggestions on how to proceed will be appreciated!

### CODE REVIEW

File: opendevin/sandbox/sandbox.py [hunk 1]
```diff
@@ -240,18 +267,35 @@ def restart_docker_container(self):
...
-                volumes={self.workspace_dir: {"bind": "/workspace", "mode": "rw"}},
+                # detach=True,
+                volumes=["/workspace"],
+                host_config=self.docker_client.create_host_config(
+                    binds={self.workspace_dir: {"bind": "/workspace", "mode": "rw"},},
+                    network_mode="host",
+                ),
             )
+            self._attach_sockets()
+            self.stdin_stream.write("echo 'hello'\n".encode("utf-8"))
```
Comment (rbren) [inline opendevin/sandbox/sandbox.py:293]: should we remove this line?

**Review by rbren — COMMENTED** [review]
This looks powerful! Will try it out 🦾

**Comment by xingyaoww** [comment]
A potentially better implementation at https://github.com/OpenDevin/OpenDevin/pull/847

**Comment by xingyaoww** [comment]
Close due to the merge of #847

**Pull request closed without merging**